ignite/contrib/handlers/lr_finder.py

```diff
@@ -0,0 +1,425 @@
+        smooth_f (float): loss smoothing factor in range [0, 1)
+        diverge_th (float): Used for stopping the search when
+            `current loss > diverge_th * best_loss`
+        memory_cache (bool): if this flag is set to True, `state_dict` of model
+            and optimizer will be cached in memory. Otherwise, they will be
+            saved to files under the `cache_dir`.
+        cache_dir (str): path for storing temporary files. If no path is
+            specified, system-wide temporary directory is used. Notice that this
+             parameter will be ignored if `memory_cache` is True.
+
+    Examples:
+
+        .. code-block:: python
+
+            from ignite.contrib.handlers import LRFinder
+
+            # Create a lr_finder
+            lr_finder = LRFinder(model, optimizer)
+
+            # Attach the lr_finder to the trainer
+            lr_finder.attach(trainer)
+
+            # Run trainer
+            trainer.run(dataloader)
+
+            # Detach lr_finder
+            lr_finder.detach()
+
+            # Get lr_finder results
+            lr_finder.get_results()
+
+            # Plot lr_finder results (requires matplotlib)
+            lr_finder.plot()
+
+            # get lr_finder suggestion for lr
+            lr_finder.lr_suggestion()
+
+        It is recommended to use the lr_finder.attach as context manager:
+
+        .. code-block:: python
+
+            from ignite.contrib.handlers import LRFinder
+
+            lr_finder = LRFinder(model, optimizer)
+
+            with lr_finder.attach(trainer):
+                trainer.run(dataloader)
+
+            lr_finder.plot()
+
+
+    References:
+        Cyclical Learning Rates for Training Neural Networks:
+        https://arxiv.org/abs/1506.01186
+
+        fastai/lr_find: https://github.com/fastai/fastai
+    """
+
+    def __init__(self, model, optimizer, num_iter=None, output_transform=lambda output: output, end_lr=10,
+                 step_mode="exp", smooth_f=0.05, diverge_th=5, memory_cache=True, cache_dir=None):
+
+        if smooth_f < 0 or smooth_f >= 1:
+            raise ValueError("smooth_f is outside the range [0, 1]")
+        if diverge_th < 1:
+            raise ValueError("diverge_th should be larger than 1")
+        if step_mode not in ["exp", "linear"]:
+            raise ValueError("expected one of (exp, linear), got {}".format(step_mode))
+        if isinstance(num_iter, int) and num_iter <= 0:
+            raise ValueError("if provided, num_iter should be a poitive int, got {}".format(num_iter))
+
+        self._model = model
+        self._optimizer = optimizer
+        self._output_transform = output_transform
+        self.num_iter = num_iter
+        self._end_lr = end_lr
+        self._step_mode = step_mode
+        self._smooth_f = smooth_f
+        self._diverge_th = diverge_th
+        self._memory_cache = memory_cache
+        self._cache_dir = cache_dir
+
+        self._engine = None
```

Comment:
Sure, it can be remove for now.

Reply:
Actually, detach can be purely internal method, IMO. Once, attached we can add detach on `Events.COMPLETED` and execute it automatically in all cases (context manager or not).

Reply:
OK, no problem, I will change detach to be purely internal. If we do it than the context manger won't really do anything, so it can be removed as well. sounds ok?

Reply:
@ItamarWilf I thought more about this, and maybe should keep only context manager API, such that it is more clear for the user that all attachments are removed outside the context. Otherwise in the following code 
```python
# Attach the lr_finder to the trainer
lr_finder.attach(trainer)
# Run trainer
trainer.run(dataloader)
```
it is not clear that all added handlers by lr finder are removed after the run. Thoughts ?

Reply:
Sounds good, "Explicit is better than implicit" :)
I will keep detach as an external method and remove the automatic detaching from `Event.COMPLETED`

Reply:
Actually, I thought about providing the API only with context manager and without `detach`
```python
with lr_finder.attach(trainer):
    trainer.run(dataloader)
```

Reply:
OK, I will try to find a way to enforce usage of the `attach` method only from within a context manager.

Reply:
Well, it seems there isn't a clean way to enforce usage of `attach` with a `with` statement. 
I suggest using `attach` just to clarify to which engine, model and optimizer the lr_finder is connected, while all the actual handler attaching and detaching happens inside the `with` statement. For example:
```python
lr_finder.attach(dummy_engine, model, optimizer)
with lr_finder:
    dummy_engine.run(dataloader)
```
This way we can keep `detach` as an internal method and we enforce the handler clean up and  model + optimizer reset.

Thoughts?

Reply:
@ItamarWilf thanks for the proposition, let me think more about the API and we'll decide which is more convenient...